### ISSUE TITLE
Is there any way to sync fewer merge commits when extracting a subfolder from a large repo?

### ISSUE THREAD

**Issue by RalfJung** [josh-project/josh#1328]
I've asked this before in https://github.com/josh-project/josh/issues/952 and was told "no", but maybe it's worth asking again... as other projects that have a presence in the Rust compiler monorepo are considering josh, we're seeing initial josh syncs that [add more than 10k commits from the parent repo](https://github.com/flip1995/rust-clippy/pull/8), meaning that about 1/3 of the commits in that history are not actually from the subproject. In Miri we have accumulated at least around 3500 of these commits (it's hard to reliably find them all so this is a lower bound), which is more than a quarter of the commits in Miri. rust-analyzer seems to be doing better, "only" getting around 1500 commits in the [initial sync](https://github.com/rust-lang/rust-analyzer/pull/17152), which is around 5% of the total commits in that history.

> When adding an external repo as a subdirectory into a monorepo (what you want to do) Josh guarantees that splitting that subdirectory back out will yield the exact same sha1's like the original repo. (Most, if not all, other filtering tools do not make that guarantee)

These commits do not originate from the subrepo, so we don't need them to be preserved. But of course josh has no way of knowing that... maybe if we could tell it which part of the history is originally from the subrepo ("this is the subrepo HEAD, everything above this, if it exists in the parent repo, must be extracted perfectly"), it could be more "sloppy" on the remaining history? But I can see how that could be anything between tricky and complete nonsense...

Cc @flip1995 @lnicola

**Comment by christian-schilling** [comment]
Interesting 🤔 

"It's hard to reliably find them" also means: "it's hard to implement a rule to skip them".

But I have been looking again at the history you posted in #952.
A lot of the merges look degenerate. the second parent is an ancestor of the first parent. I wonder if that would be a rule to use to filter them out.
Or maybe: If the second parent is an ancestor of the first *and* the diff with the first parent is empty, skip the commit.

The current rule is to keep a merge if the diff with either parent is non-empty. And to never change the number of parents a commit has.

**Comment by RalfJung** [comment]
> "It's hard to reliably find them" also means: "it's hard to implement a rule to skip them".

Yeah, I understand. And obviously keeping the round-trips working is very important.

Even if you found a better heuristic now, I have no idea how we'd migrate to that given that we've already integrated all these merges into Miri's history. Though maybe it would be worth a single force-push to Miri...

**Comment by christian-schilling** [comment]
> Even if you found a better heuristic now, I have no idea how we'd migrate to that given that we've already integrated all these merges into Miri's history. Though maybe it would be worth a single force-push to Miri...

This could also be implemented in a new filter, or as an option for the `:linear` filter.

**Comment by flip1995** [comment]
I'm still a bit hesitant with moving Clippy over to Josh, because of this comment:

> I have no idea how we'd migrate to that given that we've already integrated all these merges into Miri's history. Though maybe it would be worth a single force-push to Miri...

If this should be implemented, is there a way to move from the current `:rev` filter we're using to the new implementation without force pushing? If not, are you planning to work on this issue? Is there anything I can help with here?

If this is not planned to be implemented soon, I'll probably move to Josh anyway with the `:rev` filter, as I really want to get rid of `git subtree` :sweat_smile:

**Comment by christian-schilling** [comment]
Sorry for the late response...
So to answer the first question: If the behaviour is changed and you want to migrate, a force push will be necessary.
As the current behaviour is already relied upon, the new one (if my idea even work, we don't know that for sure yet) will be opt-in and supported alongside the current one.
Regarding planned work on it: I think I could look into it sometime next week.

**Comment by flip1995** [comment]
Thanks for looking into this! I can definitely still wait a week or two or longer, if your idea works out but you need more time implementing it.

**Comment by flip1995** [comment]
@christian-schilling Did you get to looking into this?

**Comment by tgross35** [comment]
I'm assuming most people in this thread already understand the big picture here, but I didn't 🙂 so I figure there's no harm in writing out what I think is going on with slightly more understandable graphs...

Looking at https://github.com/rust-lang/compiler-builtins/pull/946 for example[^ex], which is a sync to that repo from rust-lang/rust. The breakdown is:

* 9 non-merge commits with changes (the patches authored by folkertdev or myself)
* 2 non-merge commits related to the sync ("Update the upstream Rust version" and "Work around out-of-tree...")
* 1 merge commit related for the sync ("Merge ref 'd087f112b7d1:/library/compiler-builtins'")
* _41_ merge commits unrelated to the sync

The way that one of the non-merge (desired) commits would have gotten into the monorepo tree is via a rollup[^rollup], e.g. "use #[naked] for __rust_probestack" in rollup https://github.com/rust-lang/rust/pull/142299. With a subtree-affecting rollup pushed to master, our rust-lang/rust tree looks like this:

```text
newmaster
    |
auto merge
    |     \
    |   rollup merge 3
    |      |    \
    |      |    patch3
    |      |     |
    |      |    ...
    |      |
    |   rollup merge 2
    |      |    \
    |      |    patch2 (only commit that affects the subtree)
    |      |               \
    |   rollup merge 1     |
    |      |    \          |
    |      |    patch1     |
    |      |     |         |
  others   |    ...        |
    |      |               |
    |      /               |
rollup branch point        |
    |                      |
    |     /----------------/
patch2 branch point
    |
oldmaster
```

(The signal-to-noise ratio is much worse than it needs to be in our graph because bors doesn't know how to do merges with more than two parents, but that's besides the point.)

All merge commits from that structure seem to wind up in the monorepo->subtree sync. If I'm understanding correctly, I think this could be explained by Josh needing to "resolve" the entire graph from newmaster->patch2->oldmaster. "Resolve" meaning that if it sees a multi-parent (merge) commit in that path, then that the commit will appear in the filter, and it will trace each parent back to `oldmaster` applying the same logic. So when Josh sees that patch2 made it to `newmaster` via "auto merge", it:

* Includes "auto merge" in the subtree
* Starts checking each parent of "auto merge"
* Sees "rollup merge 3", so includes that
    * There is a clean path back to (some) master with no merges, nothing else to do here ("rollup merge 3" -> "patch3" -> "patch3 branch point")
* Sees "rollup merge 2", so includes that
    * There is a clean path back to (some) master with no merges, nothing else to do here ("rollup merge 2" -> "patch2" -> "patch2 branch point")
* etc.

Until it makes it back to `oldmaster` via every possible branch.

There is also some tricky interaction with parents: whatever "patch branch point" is appears to get included in the filtered graph[^parent]. For an example of this see https://github.com/rust-lang/miri/pull/4436: 773644b, d79d0ca, and 64eae4e are included in the sync even though they didn't bring anything new into the subtree, but they are the fork points for patches that did affect the subtree.

I think the flaw with this behavior is that even when single-parent commits that don't touch the subtree (e.g. "patch1" and "patch3") get pruned, the merges that brought these commits into `newmaster` don't get pruned. Filtered to the subtree, we should be seeing a graph like this[^resolve]:

```text
# Filtered
newmaster
    |
auto merge
    |     \
    |   rollup merge 2
    |      |    \
    |      |    patch2 (only commit that affects the subtree)
    |      |              |
    |     /---------------/
oldmaster
```

IMO this would make more sense for a default than the current behavior, but it sounds like that can't easily be changed.

[^ex]: Similar example from the miri subtree https://github.com/rust-lang/miri/pull/4431
[^rollup]: (merge a bunch of PRs to a branch `rollup-abcd` then test+merge those together
[^parent]: Possibly only if it's a merge commit? I am unsure since top-of-tree is always a merge so that's effectively always where branches happen.
[^resolve]: Obviously it can't drop the merge commits if they are nonempty, I assume `:linear` handles this somehow too

**Comment by christian-schilling** [comment]
I have been looking into this again. The reason those merges are kept is that while they don't differ from the first parent (the diff that most UIs show) they do differ from the second parent, and Josh will keep all commits that differ from any parent.
Changing this behaviour would cause all sorts of issues, most importantly breaking existing users. So I implemented a new filter that can be applied to the end of the chain to prune those merges that don't differ from the first parent `:prune=trivial-merge` in #1483 .
This is work in progress so don't push the output anywhere yet...
I tried it on the miri history and it looks good as far as I can tell.

**Comment by christian-schilling** [comment]
@flip1995 
I tried this out with clippy `:/src/tools/clippy` vs `:/src/tools/clippy:prune=trivial-merge` (applied to rust-lang/rust)

```
% git rev-list --count refs/heads/clippy-before
15305
% git rev-list --count refs/heads/clippy-after 
3690
```

Would be good to have another set of eyes on this to check the history looks as expected.

@RalfJung I'm looking into a way to allow using this for existing users (like miri) without needing a force push by "switching filters" at a certain point in the history (like `:rev()` does), but it's harder than expected because `:rev()`  allows selecting commit "older than X" but now we'd need "newer than"...

**Comment by tgross35** [comment]
I’d be happy to try it out for compiler-builtins as one of the first repos once you’re reasonably confident it’s in a near-usable state. No need for the “since” filter in this case: IMO substantially improving the signal-to-noise ratio of existing history is very well worth a force push. (Probably the same for our other smaller repos. Miri likely has different requirements here.)

**Comment by christian-schilling** [comment]
@tgross35 I think this is already usable and will merge it as is if none of you report any nasty surprises.
I checked compiler-buildins as well. Looks good to me. 

<img width="1190" height="1121" alt="Image" src="https://github.com/user-attachments/assets/79d92915-cd7e-49f3-8abd-4a31621b1106" />
<img width="1174" height="1120" alt="Image" src="https://github.com/user-attachments/assets/f97a4b0b-d71b-4305-ba71-6d41575492a2" />

I just want to be extra sure with changes like this because anything that permanently affects sha's is hard to fix once adopted.

**Comment by flip1995** [comment]
> Would be good to have another set of eyes on this to check the history looks as expected.

I'm getting _more_ commits when using the `:prune=trivial-merges` filter. Maybe I'm doing something wrong though. 

This is the [filter](https://github.com/flip1995/rust-clippy/blob/19c34d40d2b99eb692b09afe72aab9a6e94ace4a/clippy_dev/src/sync.rs#L9-L10) I'm using. The SHA in there is the latest common SHA between the Clippy and Rust repo, aka the last `git subtree` sync. When running this sync script (`cargo dev sync pull`) with the `:prune` filter, I'm getting [This branch is 27197 commits ahead of, 9 commits behind `rust-lang/rust-clippy:master`](https://github.com/flip1995/rust-clippy/tree/josh-rustup). But without the `:prune` filter, it's "only" [This branch is 15311 commits ahead of, 9 commits behind `rust-lang/rust-clippy:master`](https://github.com/flip1995/rust-clippy/tree/josh-rustup-wo).

I hope, I'm doing something wrong.

**Comment by christian-schilling** [comment]
Pruning trivial merges from just the subtree part (starting at the sha from your script)
```
% git update-ref refs/heads/clippy-subtree 20ce69b9a63bcd2756cd906fe0964d1e901e042a
% josh-filter :prune=trivial-merge refs/heads/clippy-subtree --update refs/heads/clippy-subtree-pruned
% git rev-list --count refs/heads/clippy-subtree                                                      
23755
% git rev-list --count refs/heads/clippy-subtree-pruned
23617
```
The problem is that there are already trivial merges in that old history and those get pruned as well.
That means that the pruned history, even though it does have less commits in it, it has less common commits
with the existing history and thats why github reports more commits "ahead".
The only ways I can think of to fix this are either accepting force push to clippy or waiting for the "since filter" that I suggested above for miri already.

**Comment by flip1995** [comment]
If this could be "fixed" with a single force push once, I'd be fine with that. But if the ETA of the "since filter" is rather weeks/months than years, I would rather wait for that.

**Comment by christian-schilling** [comment]
So, I got this to work in #1519 . Was quite a puzzle and surprisingly simple in the end...

`:rev(20ce69b9a63bcd2756cd906fe0964d1e901e042a:prefix=src/tools/clippy):/src/tools/clippy:from(b19acfca58b1f2cc7b4fe1adadd3a8b5c97b840d:prune=trivial-merge)`

Gives something like this:

<img width="751" height="243" alt="Image" src="https://github.com/user-attachments/assets/f74015b3-0754-47bd-8ae7-ae722f0b5c8e" />

Note that the sha given is not from the original history but from the history already filtered with `:rev(20ce69b9a63bcd2756cd906fe0964d1e901e042a:prefix=src/tools/clippy):/src/tools/clippy`

I choose an arbitary recent commit to try this out.

**Comment by christian-schilling** [comment]
@flip1995 Does this work for your case?

**Comment by flip1995** [comment]
Sorry for not responding for so long. I wanted to try it right after the sync to try it in the same conditions as when I would switch over to JOSH.

I'm still getting [+15425 commits](https://github.com/flip1995/rust-clippy/tree/rustup-from). This is what I was doing:

```
git fetch http://localhost:42042/rust-lang/rust.git@f46475914de626785090a05ae037578aaa119fc8:rev(d9fb15c4b1ebe9e7dc419e07f53af681d7860cbe:prefix=src/tools/clippy):/src/tools/clippy:from(2a5c4ae5ed0566c19267dd319b904daf1e3dbaf9:prune=trivial-merge).git
git merge FETCH_HEAD --no-verify --no-ff -m "Merge from rustc"
```

where 

- `f46475914de626785090a05ae037578aaa119fc8` is the commit of the latest nightly in the rust-lang/rust repo (2025-10-18)
- `d9fb15c4b1ebe9e7dc419e07f53af681d7860cbe` is the latest commit that was synced from the Clippy repo to the Rust repo with `git subtree`
- `2a5c4ae5ed0566c19267dd319b904daf1e3dbaf9` is the commit from `git rev-parse FETCH_HEAD` with the filter `:rev(d9fb15c4b1ebe9e7dc419e07f53af681d7860cbe:prefix=src/tools/clippy):/src/tools/clippy` applied.


Anything I'm doing wrong?

**Comment by christian-schilling** [comment]
@flip1995 I don't think you are doing anything wrong.
After looking at crazy commits graphs for 3 nights already...

![Image](https://github.com/user-attachments/assets/df5c4f33-98c8-4e1e-a93a-2e836d7acf7f)

... I can say all the git clients are really not working well with such big and complex history, I keep discovering bugs and limitations in those (I tried a few already...)

Now I found this: https://github.com/gitgitgadget/git/pull/493
It says that even with this patched `git subtree` it still might get tripped up. I'm hoping that we are looking at a bug/tripup in `git subtree` here and josh gives the correct output, but so far I have not found the proof for that. So I keep looking since if this is a josh bug I'd like to know...

Ah: And this issue has nothing to do with the trivial-merge pruning...I think..

**Comment by RalfJung** [comment]
> Now I found this: https://github.com/gitgitgadget/git/pull/493

Yeah there's a reason we went to JOSH with Miri -- we simply couldn't get any of the other tools to even work.^^
Kudos on being the only tool that can cope with this. :)

**Comment by flip1995** [comment]
Oh, _crazy_ is an understatement. If there's a way to solve this with a force push to the Clippy `master` branch, I'm happy to do that. But I don't know how I would even start to clean this up 😅 

Yeah, that's the branch of `git subtree` we are using. It's the only thing that worked for Clippy.

**Comment by christian-schilling** [comment]
After a lot more digging I found *an* explanation for differences in the git subtree vs josh subtree:
https://github.com/search?q=org%3Arust-lang+%22Rollup+merge+of+%23124957+-+compiler-errors%3Abuiltin-deref%2C+r%3Dmichaelwoerister%22&type=commits

This is the original: https://github.com/rust-lang/rust/commit/9a9ec90567f20a545fdb3366c7721d6472786b91
The others are replicated versions of it. What is interesting is that there are more replicas of the same commit inside the rust-lang/rust repo: They are the commits that have been exported to the subtree repos and then re-imported into rust-lang/rust.
I'm not sure about subtree, but Josh sould usually be able to avoid that duplication because it searches for the original commits in the target branch when doing the "push".

However....this is where everything breaks:

```
% git cat-file commit 9a9ec90567f20a545fdb3366c7721d6472786b91
tree 7446c503f85a0618ebf2ccf4a94c446d1e481667
parent aa68901e365afaa2f1026787d6b7e74507ce8c20
parent d50c2b0a52f31fbbf6e6bfa27c129b513fdddb0b
author Matthias Krüger <matthias.krueger@famsik.de> 1715350247 +0200
committer GitHub <noreply@github.com> 1715350247 +0200
gpgsig -----BEGIN PGP SIGNATURE-----
 
 wsFcBAABCAAQBQJmPirnCRC1aQ7uu5UhlAAAPykQAK5nxzsmfigSGuXkiFLny/oH
 VuvRBSIWY2EqcphacoWFbhCqsBccl63SKfrPc3ormf9pWMysQHb0eWFjU0lMqRVj
 A9CkfNODtGpDCQIeKrDmbpep+gqNEfHtksstF7lfyEwZkvtG694zaienLiiXSWHo
 O22iFOVk8ncHAxlJSDK0R4lGPESpLVKlqru8UiAg+VmEqwR/Wk68NvhuK2Vl97sh
 XWR8nH+jJD+kg48oAorQ9pdY67hyro5cTTm7Ops8iguwMorlDFWL2jCEwIj4YDxU
 P3WdAXbN12x606diIwlvEF9NEa1JX+7QFwSulnkocoKXTVUHK6PiUiUfxjUJJp4Y
 Mcj+WVmGVfHHlbajMv7iJEjaUbRUatadkjeNaUYYEm134zqI+F99eUv911lC740t
 +YWlzg/MFm4SmoCN6NOt1b2yGC7nK7ZSkgo/F1Xg56PB+QglO1zNac6qkNQzr2qF
 nTOGGaHf+JZBJwa8rNM3FrxVmNKcIUXJZC0PejJPCCtgpZEMsaMDEEdmUdN9OhOH
 6yxMM0e1Anb8MA17X4R4WPcWFYxT1JaxLTdetgZiDVCrLE4m4iawuv+XWO1Cx1Ia
 ht3x6l97z6NzHWgDTrqlA/6RnVYWMbWk6YVCkTcJbp8RH5OCvrDQmZ4f0LVmLYFp
 RsrJWvTpj6lQBg2Zk7xO
 =5sFR
 -----END PGP SIGNATURE-----
 

Rollup merge of #124957 - compiler-errors:builtin-deref, r=michaelwoerister

Make `Ty::builtin_deref` just return a `Ty`

Nowhere in the compiler are we using the mutability part of the `TyAndMut` that we used to return.
```

It looks like `git subtree` strips signatures when exporting commits. This breaks the roundtrip.
For `git subtree` it is not that big of an issue because commit duplication is assumed anyway: The original change was made in the subtree path of the main repo, the change that comes back will be in the (less deep) path of the subtree itself.

We had rountrip issues because of signatures in the past as well, and changed Josh to not strip signatures by default. (That's when the `:unsign` filter was introduced in case the old behaviour is still needed...)
That means the signatures are present in partial commits, but of course not valid. Still this allows getting the same sha out again after inverse filtering.

Unfortunatly there are also signed commits in rust-lang/rust-clippy.
So even with the `:unsign` filter we can't get to the same shas.

All in all I'm now quite confident the divergence in history is not due to a bug in Josh, and a force push to clippy is the best solution if possible.

You won't need the `:from` filter then, just the `:rev` to get all the old history, and `:prune=trivial-merge` can be applied to the entire history.

Totally out of scope for this issue, but it would be interesting to figure out how to do signed commits in a "subtree aware" way that actually works... 🤔

**Comment by flip1995** [comment]
> All in all I'm now quite confident the divergence in history is not due to a bug in Josh, and a force push to clippy is the best solution if possible.
> 
> You won't need the `:from` filter then, just the `:rev` to get all the old history, and `:prune=trivial-merge` can be applied to the entire history.

Thanks for all the digging!

How would I go about preparing a branch, which I could then potentially force push? I'm not sure where to start doing that. Do I have to undo past `git subtree` merges? Or is it enough to just use some josh filter and then force push that result?

**Comment by RalfJung** [comment]
The filter would be `:rev(d9fb15c4b1ebe9e7dc419e07f53af681d7860cbe:prefix=src/tools/clippy):/src/tools/clippy:prune=trivial-merge`, and the idea would be to run that on the current rustc history and use the result as new clippy main branch.

**Comment by flip1995** [comment]
Ah, got it! Also tested it and this would give us this branch: https://github.com/flip1995/rust-clippy/tree/josh-force so ~+3k commits compared to master. But not doubling the number of commits. I think this is what we want then.

**Comment by christian-schilling** [comment]
I'm just wondering what to do about the existing refs other than master.
There are tags. Not sure that they will still point to ancestors of master after the force push. (they could!)
If they don't maybe it would make sense to update them...have to figure out how 🤔

**Comment by flip1995** [comment]
Good point. The tags called something like `0.212.0` we can just delete, I think. No one should rely on those and the documentation of those were deleted years ago already. The `rust-1.**.0` tags we probably will have to re-tag. But I think that we can do with a script that finds the new commits by commit message or something.

**Comment by RalfJung** [comment]
What would go wrong if we just left the tags unchanged?

**Comment by christian-schilling** [comment]
I checked: The tags would be disconnected from the main branch.
The question is what are they used for? Would anything break if they change?

The advantage of updating them would be that they would still appear in annotated git log.
So if someone for example would use bit blame or bisect for example to find a certain past commit that might have introduced a bug or something, having the tags pointing to the commits in the main branch would mean it would be easy to tell "this change was after tag X but before tag Y.

How were they created? My guess is they are exactly corresponding to the tags in r-l/r ?
In that case they would be easy to recreate: Just pull the tags through josh-proxy with the same filter and done.

Could even be an addition to the r-l/josh-sync tool: Auto pull the rust release tags (or whatever additional refs) to the subrepos if wanted.

**Comment by vlad-ivanov-name** [comment]
If something like e.g. NixOS pins revision by tag + contents hash (as opposed to commit hash), this would be noticed. Same for other package managers that pin tag and hash (not sure how common it is compared to pinning commit SHA).

**Comment by RalfJung** [comment]
The intention is for downstream to use the release tarballs, not git tags... but that may not stop them. ;)

**Comment by flip1995** [comment]
>  The question is what are they used for? Would anything break if they change?

The only use inside the Clippy repo is  to have a tag that marks the commit that got into the release and to trigger documentation generation for the release. For downstream users, I don't know how much people rely on them. 

> How were they created? My guess is they are exactly corresponding to the tags in r-l/r ?

[Manually](https://doc.rust-lang.org/clippy/development/infrastructure/release.html#find-the-clippy-commit) at the moment. They are corresponding to the last commit synced to the Rust repo for that release. Automating this with josh is definitely an option.

**Comment by flip1995** [comment]
We could also keep the old history and tags around by moving the current `master` branch to a `legacy` branch. So the hashes and tags would stay untouched. And then add branch protection to the `legacy` branch to make sure no one updates that one. And then future tags are added to the new `master` branch.

**Comment by Kobzol** [comment]
Also note that force-pushing will essentially "break" all open PRs, requiring a rebase + force-push on each one, and sometimes the rebase does not go well. So preparing some guide for contributors on how to rebase (or even cherry-pick) on top of the force-pushed master would be useful.

**Comment by RalfJung** [comment]
> The filter would be `:rev(d9fb15c4b1ebe9e7dc419e07f53af681d7860cbe:prefix=src/tools/clippy):/src/tools/clippy:prune=trivial-merge`, and the idea would be to run that on the current rustc history and use the result as new clippy main branch.

FWIW this would not fix the actual underlying issue. That issue is baked into rustc history now: we have a [commit](https://github.com/josh-project/josh/issues/1328#issuecomment-3447064487) with a signature in rustc that touched clippy, and the synced version of that commit lost the signature, and that synced commit has been integrated back into the rustc history. No amount of force-pushing to clippy can undo that. The filter above still has the duplicated history. There's just less of it to duplicate due to the trivial-merge pruning.

**Comment by RalfJung** [comment]
Also, turns out a bunch of the discussion above is somewhat off-topic for the original issue description here.^^ Many of the new commits in the josh-generated clippy history aren't actually "trivial merges", they are caused by history duplication caused by the commit described above, which is an independent problem.

**Comment by tgross35** [comment]
I finally got the chance to play around with this (inspired by the ratio of [76 merge commits to three useful commits](https://github.com/rust-lang/compiler-builtins/pull/1068)) and the recent history looks good: https://github.com/rust-lang/compiler-builtins/tree/prune-trival-merges. However, it looks like it rewrote an awful lot. Checking the merge base returns:

```text
$ git merge-base upstream/prune-trival-merges upstream/main
237809534591b2f5ea00ee84b79685858eaa81fe
$ git log -1 237809534591b2f5ea00ee84b79685858eaa81fe
commit 237809534591b2f5ea00ee84b79685858eaa81fe
Merge: 9710af9a33 827f9a8a01
Author: Alex Crichton <alex@alexcrichton.com>
Date:   Mon Jan 7 11:11:19 2019 -0600

    Merge pull request #270 from denzp/master

    Don't build compiler-rt for NVPTX
```

which is way before we started using Josh (that was in May of last year https://github.com/rust-lang/rust/pull/141229). Any idea why this would happen? I don't think any pre-Josh history should be touched, unless it is incorrectly considering something in the subtree's history to be a trivial merge.

**Comment by tgross35** [comment]
Here's the log from around that time. Left is the history with `prune=trivial-merges`, right is the correct history at `main`, the merge-base commit is kinda highlighted:

<img width="1512" height="310" alt="Image" src="https://github.com/user-attachments/assets/80280f02-2282-4917-8df6-a279077a6784" />

It looks like there was a master->branch merge at some point in history which got pruned. Which is a yucky thing to have in there, but I'd rather not rewrite all history since 2019 if it can be avoided. Is it expected that this kind of merge gets pruned, in which case I think we should use the `:from` filter with it? Or is this unexpected behavior of the pruning?

**Comment by christian-schilling** [comment]
@tgross35 This was a known limitation of the `:prune=trivial-merge` filter: It can't distinguish between merges that become trivial du to the filtering, and those that have been tivial in the original history. Thus breaking round trip.
(not so) coincidentally #1174 was merged a few days ago to address this:
https://josh-project.github.io/josh/reference/filters.html#history-option
As the result of the plan I metioned there:
[#t-infra > Prune empty merges in Josh @ 💬](https://rust-lang.zulipchat.com/#narrow/channel/242791-t-infra/topic/Prune.20empty.20merges.20in.20Josh/near/564821785)

I'm now waiting for #1785 to get merged to write up what's changed, since both are backwards incompatible changes. That's not ideal, but I think safe to do here, since I'm sure nobody used those as "stored filters" but only as externally configured ones.
So they will require updating the config for existing subtrees (miri) but then still be able to keep creating the same history as before.

So the changes:
With #1774 merges that become trivial are removed by default, while the "already trivial" ones are kept. So that would fix the issue you're describing above. It also intriduces a way to modify the filter behavior to work as is did before and keep all trivial merges as it did before. Meaning miri needs to change from
`:rev(75dd959a3a40eb5b4574f8d2e23aa6efbeb33573:prefix=src/tools/miri):/src/tools/miri`
to
`:rev(75dd959a3a40eb5b4574f8d2e23aa6efbeb33573:prefix=src/tools/miri):~(history:"keep-trivial-merges")[:/src/tools/miri]`

The `:prune=trivial-merge` should not be used normally except when the actuall pruning of all trivial merges is wanted.

With #1785 the `:rev` will become more powerful (also replacing the previous `:from` filter) to be able deal with existing subtree histories that want to switch from "old behaviour" to "new behaviour" without affecting old history and thus avoiding the need for force push (might still choose to of course..)
It also gains "exact match" for cases where a single commit in the history (or limited number) needs to the fixed somehow like we saw with the signature issue.

This will mean another config change though:
`:rev(<=75dd959a3a40eb5b4574f8d2e23aa6efbeb33573:prefix=src/tools/miri):~(history:"keep-trivial-merges")[:/src/tools/miri]`

**Comment by RalfJung** [comment]
That sounds great, thanks for the update and the hard work!

Would this more powerful rev also allow Miri to switch to pruning trivial merges from some point onwards, and what would the filter for that look like?

**Comment by christian-schilling** [comment]
```
:rev(<=75dd959a3a40eb5b4574f8d2e23aa6efbeb33573:prefix=src/tools/miri):rev(
    <SHA_TO_START_PRUNE_FROM:~(history:"keep-trivial-merges")[:/src/tools/miri]
    _:/src/tools/miri
)
```
Note that the sha in the second rev must be in the history that is output from the first.
Maybe there are other ways to write this as well 🤔
